Are HashSets not supported?

Tried to derive `Encode` and `Decode` on a struct with a `HashSet` property but getting the following error:

```
error[E0277]: the trait bound `HashSet<u32>: Encode` is not satisfied
 --> src/main.rs:6:10
  |
6 | #[derive(Encode, Decode)]
  |          ^^^^^^ the trait `Encode` is not implemented for `HashSet<u32>`
  |
  = note: this error originates in the derive macro `Encode` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `HashSet<u32>: Decode` is not satisfied
 --> src/main.rs:6:18
  |
6 | #[derive(Encode, Decode)]
  |                  ^^^^^^ the trait `Decode` is not implemented for `HashSet<u32>`
  |
  = note: this error originates in the derive macro `Decode` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```

--- a/fuzz/fuzz_targets/roundtrip.rs
+++ b/fuzz/fuzz_targets/roundtrip.rs
@@ -1,7 +1,7 @@
 #![no_main]
 use libfuzzer_sys::fuzz_target;
 
-use std::collections::{BTreeMap, BTreeSet, HashMap, VecDeque};
+use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet, VecDeque};
 use std::ffi::CString;
 use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
 use std::num::{NonZeroI128, NonZeroI32, NonZeroU128, NonZeroU32};
@@ -14,6 +14,7 @@ use std::time::{Duration, SystemTime};
 enum AllTypes {
     BTreeMap(BTreeMap<u8, u8>),
     HashMap(HashMap<u8, u8>),
+    HashSet(HashSet<u8>),
     BTreeSet(BTreeSet<u8>),
     VecDeque(VecDeque<AllTypes>),
     Vec(Vec<AllTypes>),

--- a/src/features/impl_std.rs
+++ b/src/features/impl_std.rs
@@ -6,8 +6,9 @@ use crate::{
 };
 use core::time::Duration;
 use std::{
-    collections::HashMap,
+    collections::{HashMap, HashSet},
     ffi::{CStr, CString},
+    hash::Hash,
     io::Read,
     net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6},
     path::{Path, PathBuf},
@@ -390,5 +391,38 @@ where
             map.insert(k, v);
         }
         Ok(map)
+    }
+}
+
+impl<T> Decode for HashSet<T>
+where
+    T: Decode + Eq + Hash,
+{
+    fn decode<D: Decoder>(decoder: &mut D) -> Result<Self, DecodeError> {
+        let len = crate::de::decode_slice_len(decoder)?;
+        decoder.claim_container_read::<T>(len)?;
+
+        let mut map = HashSet::new();
+        for _ in 0..len {
+            // See the documentation on `unclaim_bytes_read` as to why we're doing this here
+            decoder.unclaim_bytes_read(core::mem::size_of::<T>());
+
+            let key = T::decode(decoder)?;
+            map.insert(key);
+        }
+        Ok(map)
+    }
+}
+
+impl<T> Encode for HashSet<T>
+where
+    T: Encode,
+{
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        crate::enc::encode_slice_len(encoder, self.len())?;
+        for item in self.iter() {
+            item.encode(encoder)?;
+        }
+        Ok(())
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -17,7 +17,7 @@
 //!
 //! |Name  |Default?|Supported types for Encode/Decode|Enabled methods                                                  |Other|
 //! |------|--------|-----------------------------------------|-----------------------------------------------------------------|-----|
-//! |std   | Yes    |`HashMap`|`decode_from_std_read` and `encode_into_std_write`|
+//! |std   | Yes    |`HashMap` and `HashSet`|`decode_from_std_read` and `encode_into_std_write`|
 //! |alloc | Yes    |All common containers in alloc, like `Vec`, `String`, `Box`|`encode_to_vec`|
 //! |atomic| Yes    |All `Atomic*` integer types, e.g. `AtomicUsize`, and `AtomicBool`||
 //! |derive| Yes    |||Enables the `BorrowDecode`, `Decode` and `Encode` derive macros|

--- a/tests/alloc.rs
+++ b/tests/alloc.rs
@@ -129,6 +129,7 @@ fn test_container_limits() {
         #[cfg(feature = "std")]
         {
             validate_fail::<std::collections::HashMap<i32, i32>>(slice);
+            validate_fail::<std::collections::HashSet<i32>>(slice);
         }
     }
 }

--- a/tests/std.rs
+++ b/tests/std.rs
@@ -101,6 +101,11 @@ fn test_std_commons() {
     map.insert("you".to_owned(), "doing?".to_owned());
     the_same(map);
 
+    let mut set = std::collections::HashSet::new();
+    set.insert("Hello".to_string());
+    set.insert("World".to_string());
+    the_same(set);
+
     // Borrowed values
     let config = bincode::config::standard();
     let mut buffer = [0u8; 1024];

